### PR TITLE
Converts the Skill Tree Tables file into YAML

### PR DESCRIPTION
**Pull request by Atemo**
* **Addressed Issue(s)**: N/A

* **Server Mode**: Both

* **Description of Pull Request**: 

  * Converts the Skill Tree Tables file into YAML.
  * Includes CSV2YAML converter.

----------------------------------------------------------------------------------------------
Note: the skills tree have been verified after loading vs the txt data, then adjusted according to the skill tree client side (SkillInfoList.lua, adjusted on renewal only).

### CODE REVIEW

**Comment by Lemongrass3110** [comment]
Seems like CI is crashing during or after skill db loading:
```
[Status]: Done reading '151' entries in 'db/re/skill_tree.yml'
Command exited with code -1073741819
```

File: src/tool/csv2yaml.cpp [hunk 1]
```diff
@@ -4115,3 +4137,111 @@ static bool read_constdb(char* fields[], int columns, int current) {
...
+			if (subit.baselv > 0)
+				body << YAML::Key << "BaseLevel" << YAML::Value << subit.baselv;
+			if (subit.joblv > 0)
+				body << YAML::Key << "JobLevel" << YAML::Value << subit.joblv;
+			body << YAML::EndMap;
+		}
+		body << YAML::EndSeq;
+		body << YAML::EndMap;
+		return true;
+	}
```
Comment (aleos89) [inline src/tool/csv2yaml.cpp:4246]:
```suggestion
	}
}
```

Missing a final curly bracket.

**Review by aleos89 — COMMENTED** [review]
Just a few more minor things!

File: src/map/mob.cpp [hunk 2]
```diff
@@ -4015,8 +4015,21 @@ int mob_clone_spawn(struct map_session_data *sd, int16 m, int16 x, int16 y, cons
 	sd->fd = 0;
 
 	//Go Backwards to give better priority to advanced skills.
-	for (i=0,j = MAX_SKILL_TREE-1;j>=0 && i< MAX_MOBSKILL ;j--) {
-		uint16 skill_id = skill_tree[pc_class2idx(sd->status.class_)][j].skill_id;
+	std::shared_ptr<s_skill_tree> tree = skill_tree_db.find(sd->status.class_);
+
+	if (tree == nullptr || tree->skills.empty())
+		return 0;
```
Comment (Lemongrass3110) [inline src/map/mob.cpp:4021]: This "return" cannot be done like this.
It is okay if a clonse does not have skills.
Also if you return here, the "fd" of the cloned player will be lost.

File: src/map/mob.cpp [hunk 3]
```diff
@@ -4015,8 +4015,21 @@ int mob_clone_spawn(struct map_session_data *sd, int16 m, int16 x, int16 y, cons
...
+
+	std::vector<uint16> skill_list;
+
+	for (const auto &it : tree->skills)
+		skill_list.push_back(it.first);
+	std::sort(skill_list.rbegin(), skill_list.rend());
+
+	for (const auto &it : skill_list) {
+		if (i >= MAX_MOBSKILL)
+			break;
```
Comment (Lemongrass3110) [inline src/map/mob.cpp:4031]: "i" could be removed by checking db->skill.size()

File: src/map/pc.cpp [hunk 4]
```diff
@@ -12073,101 +12101,299 @@ int pc_split_atoui(char* str, unsigned int* val, char sep, int max)
...
-	uint16 skill_id, skill_lv, skill_lv_max;
-	int idx, class_;
-	unsigned int i, offset, skill_idx;
 
-	class_  = atoi(fields[0]);
-	skill_id = (uint16)atoi(fields[1]);
-	skill_lv = (uint16)atoi(fields[2]);
+std::shared_ptr<s_skill_tree_entry> SkillTreeDatabase::get_skill_data(int class_, uint16 skill_id) {
+	std::shared_ptr<s_skill_tree> tree = this->find(class_);
+	std::shared_ptr<s_skill_tree_entry> entry = nullptr;
```
Comment (Lemongrass3110) [inline src/map/pc.cpp:12107]:
```suggestion
```

File: src/map/pc.cpp [hunk 5]
```diff
@@ -12073,101 +12101,299 @@ int pc_split_atoui(char* str, unsigned int* val, char sep, int max)
...
-	else if (columns == 3 + MAX_PC_SKILL_REQUIRE * 2) {
-		baselv = joblv = 0;
-		offset = 3;
-	}
-	else {
-		ShowWarning("pc_readdb_skilltree: Invalid number of colums in skill %hu of job %d's tree.\n", skill_id, class_);
-		return false;
-	}
+	if (tree != nullptr)
+		entry = util::umap_find(tree->skills, skill_id);
```
Comment (Lemongrass3110) [inline src/map/pc.cpp:12110]:
```suggestion
		return util::umap_find(tree->skills, skill_id);
```

File: src/map/pc.cpp [hunk 6]
```diff
@@ -12073,101 +12101,299 @@ int pc_split_atoui(char* str, unsigned int* val, char sep, int max)
...
 
-	if(!pcdb_checkid(class_))
-	{
-		ShowWarning("pc_readdb_skilltree: Invalid job class %d specified.\n", class_);
-		return false;
-	}
-	idx = pc_class2idx(class_);
+	return entry;
```
Comment (Lemongrass3110) [inline src/map/pc.cpp:12112]:
```suggestion
	return nullptr;
```

File: src/map/pc.cpp [hunk 7]
```diff
@@ -12073,101 +12101,299 @@ int pc_split_atoui(char* str, unsigned int* val, char sep, int max)
...
+
+			if (max_lv > MAX_SKILL_LEVEL) {
+				this->invalidWarning(it["MaxLevel"], "MaxLevel exceeds the maximum skill level of %d, skipping.\n", MAX_SKILL_LEVEL);
+				return 0;
+			}
+
+			uint16 skill_lv_max = skill_get_max(skill_id);
+
+			if (max_lv > skill_lv_max) {
+				this->invalidWarning(it["MaxLevel"], "Skill %s's level %hu exceeds job %s's max level %hu. Capping skill level.\n", skill_name.c_str(), max_lv, job_name.c_str(), skill_lv_max);
```
Comment (Lemongrass3110) [inline src/map/pc.cpp:12216]: Here the message should be changed?

File: src/map/pc.cpp [hunk 8]
```diff
@@ -12073,101 +12101,299 @@ int pc_split_atoui(char* str, unsigned int* val, char sep, int max)
...
 
-		if (skill_id == 0) {
-			if (skill_tree[idx][skill_idx].need[i].skill_id > 0) { // Remove pre-requisite
-				skill_tree[idx][skill_idx].need[i].skill_id = 0;
-				skill_tree[idx][skill_idx].need[i].skill_lv = 0;
+					if (lv_req > MAX_SKILL_LEVEL) {
+						this->invalidWarning(Requiresit["Level"], "Level exceeds the maximum skill level of %d, skipping.\n", MAX_SKILL_LEVEL);
+						return 0;
+					}
+
```
Comment (Lemongrass3110) [inline src/map/pc.cpp:12292]: Here just like above it should check for actual maximum level too?

File: src/map/pc.cpp [hunk 9]
```diff
@@ -12073,101 +12101,299 @@ int pc_split_atoui(char* str, unsigned int* val, char sep, int max)
...
-			skill_lv = skill_lv_max;
+	}
+
+	// remove skills with max_lv = 0
+	std::unordered_map<uint16, std::vector<uint16>> max_lv_zero;
+
+	for (const auto &job : *this) {
+		if (job.second->skills.empty())
+			continue;
+		for (const auto &skill : job.second->skills) {
```
Comment (Lemongrass3110) [inline src/map/pc.cpp:12385]: Why not use iterator here?
Then it could be removed in one loop instead of having to store it in another map?

**Review by Lemongrass3110 — COMMENTED** [review]
Looks good overall. Just a few small things.

**Review by Lemongrass3110 — APPROVED** [review]

**Review by aleos89 — APPROVED** [review]

**Pull request merged**